src/parser/deserializer.ts

```diff
@@ -152,6 +156,9 @@ function deserializeObject(
     // If are at the end of the buffer there is a problem with the document
     if (i >= buffer.byteLength) throw new Error('Bad BSON Document: illegal CString');
     const name = isArray ? arrayIndex++ : buffer.toString('utf8', index, i);
+    if (isPossibleDBRef !== false && (name as string)[0] === '$') {
+      isPossibleDBRef = allowedDBRefKeys.test(name as string);
+    }
```

Comment:
I guess the casting is unavoidable here, maybe if we checked `isArray` it would narrow? but not worth spending too much time on.